File: ValidParenthesis.py
class Solution:
    def isValid(self, s: str) -> bool:
        open_brackets = ['(', '{', '[']
        close_brackets = [')', '}', ']']
        stack = []

        i = 0
        while i < len(s):
            while i < len(s) and s[i] in open_brackets:
                stack.append(s[i])
                i += 1
            if len(stack) == 0:
                return False
            while i < len(s) and stack:
                b = stack.pop()
                try:
                    bracket_index = open_brackets.index(b)
                except ValueError:
                    return False
                print(s[i], " ", close_brackets[bracket_index])
                if s[i] != close_brackets[bracket_index]:
                    return False
                i += 1
        if stack:
            return False
        return True

File: ValidParenthesis.cpp
#include <unordered_map>
#include <stack>
#include <string>
#include <iostream>

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        unordered_map<char,char> match_for_c = {
            {')', '('},
            {'}', '{'},
            {']', '['}
        };
        stack<char> open_stack;
        string open_parens = "({[";
        for (char c : s)
        {
            if (open_parens.find(c) != string::npos)
            {
                open_stack.push(c);
                continue;
            }
            if (open_stack.empty() || open_stack.top() != match_for_c[c])
            {
                return false;
            }
            open_stack.pop();
        }
        return open_stack.empty();
    }
};

int main()
{
    Solution sol;
    cout << sol.isValid("[]");
    return 0;
}